File: ms-passwdcheck/passwdcheck.js
//
//
//
//
//
//
// Source URLs:
// https://www.microsoft.com/security/pc-security/password-checker.aspx
// https://www.microsoft.com/security/pc-security/assets/scripts/passwdcheck.js
//
//
//
//
//
//

/* New Password checker script*/
var alpha = "abcdefghijklmnopqrstuvwxyz";
var upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
var upper_punct = "~`!@#$%^&*()-_+=";
var digits = "1234567890";

var totalChars = 0x7f - 0x20;
var alphaChars = alpha.length;
var upperChars = upper.length;
var upper_punctChars = upper_punct.length;
var digitChars = digits.length;
var otherChars = totalChars - (alphaChars + upperChars + upper_punctChars + digitChars);

function GEId(sID) {
    return document.getElementById(sID);
}
function calculateBits(passWord) {
    
    if (passWord.length < 0) {
        //alert("Please supply the password");
        return 0;
    }

    var fAlpha = false;
    var fUpper = false;
    var fUpperPunct = false;
    var fDigit = false;
    var fOther = false;
    var charset = 0;

    //var c=new Array();

    for (var i = 0; i < passWord.length; i++) {
        var char = passWord.charAt(i);

        if (alpha.indexOf(char) != -1)
            fAlpha = true;
        else if (upper.indexOf(char) != -1)
            fUpper = true;
        else if (digits.indexOf(char) != -1)
            fDigit = true;
        else if (upper_punct.indexOf(char) != -1)
            fUpperPunct = true;
        else
            fOther = true;

    }

   
    if (fAlpha)
        charset += alphaChars;
    if (fUpper)
        charset += upperChars;
    if (fDigit)
        charset += digitChars;
    if (fUpperPunct)
        charset += upper_punctChars;
    if (fOther)
        charset += otherChars;

    var bits = Math.log(charset) * (passWord.length / Math.log(2));
    
    //alert(Math.floor(bits));
    return Math.floor(bits);
}

function DispPwdStrength(iN, sHL) {
    if (iN > 4) {
        iN = 4;
    }
    for (var i = 0; i < 5; i++) {
        var sHCR = "pwdChkCon0"; if (i <= iN) {
            sHCR = sHL;
        } if (i > 0) {
            GEId("idSM" + i).className = sHCR;
        }
        GEId("idSMT" + i).style.display = ((i == iN) ? "inline" : "none");
    }
}

function EvalPwdStrength(oF, sP) {
    //alert("start");
    var bits = calculateBits(sP);
    //alert(bits);
    if (bits >= 128) {
        DispPwdStrength(4, 'pwdChkCon4');
    }
    else if (bits < 128 && bits >= 64) {
        DispPwdStrength(3, 'pwdChkCon3');
    }
    else if (bits<64 && bits>=56) {
        DispPwdStrength(2, 'pwdChkCon2');
    }
    else if (bits<56) {
        DispPwdStrength(1, 'pwdChkCon1');
    }
    else {
        DispPwdStrength(0, 'pwdChkCon0');
    }
}

File: ms-passwdcheck/passwdcheck.py
from math import floor, log

# Python port of:
# https://www.microsoft.com/security/pc-security/assets/scripts/passwdcheck.js
#
# Main Page:
# https://www.microsoft.com/security/pc-security/password-checker.aspx
###############################################################################

__version__ = "1.0"
__author__ = "nomuus"
__copyright__ = "Microsoft"
__email__ = "mu*nre*txemusu*da"[::-1].replace('*', '') + "!@#$%^&*()"[1] + "nomuus" + "+..com"[2:]
__company__ = "www.nomuus.com"
__description__ = "Python Port of Microsoft Password Checker"

###############################################################################

alpha = "abcdefghijklmnopqrstuvwxyz"
upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
punct = "~`!@#$%^&*()-_+="
digit = "1234567890"
total_chars = 0x7f - 0x20
alpha_chars = len(alpha)
upper_chars = len(upper)
digit_chars = len(digit)
upper_punct = len(punct)
other_chars = total_chars - (alpha_chars + upper_chars + upper_punct + digit_chars)
strength_labels = ["None", "Weak", "Medium", "Strong", "Best"]

###############################################################################

def calculate_bits(password):
	len_pwd = len(password)
	if len_pwd < 1:
		return 0
	
	b_alpha = False
	b_upper = False
	b_digit = False
	b_punct = False
	b_other = False
	charset = 0
	
	for c in password:
		if alpha.find(c) != -1:
			b_alpha = True
		elif upper.find(c) != -1:
			b_upper = True
		elif digit.find(c) != -1:
			b_digit = True
		elif punct.find(c) != -1:
			b_punct = True
		else:
			b_other = True
	
	if b_alpha:
		charset += alpha_chars
	if b_upper:
		charset += upper_chars
	if b_digit:
		charset += digit_chars
	if b_punct:
		charset += upper_punct
	if b_other:
		charset += other_chars

	bits = log(charset) * (len_pwd / log(2))
	return floor(bits)
	
###############################################################################
	
def eval_strength(password):
	bits = calculate_bits(password)
	if bits >= 128:
		display_strength(4)
	elif bits < 128 and bits >= 64:
		display_strength(3)
	elif bits < 64 and bits >= 56:
		display_strength(2)
	elif bits < 56:
		display_strength(1)
	else:
		display_strength(0)

###############################################################################

def display_strength(index):
	print strength_labels[index]
	
###############################################################################
###############################################################################

if __name__ == "__main__":
	x = raw_input("Enter a password: ")
	eval_strength(x)
